src/metabase/query_processor/parameters.clj

```diff
@@ -10,18 +10,116 @@
+    :int-value [[group-label (Integer/parseInt group-value)]]
+    (:date :date-1 :date-2) [[group-label (parse-absolute-date group-value)]]
+    [[group-label group-value]]))
+
+(defn ^:private regex->parser
+  "Takes a regex and labels matching the regex capturing groups. Returns a parser which
+  takes a parameter value, validates the value against regex and gives a map of labels
+  and group values. Respects the following special label names:
+      :unit – finds a matching date unit and merges date unit operations to the result
+      :int-value – converts the group value to integer
+      :date[-index] – converts the group value to absolute date"
+  [regex, group-labels]
+  (fn [param-value]
+    (when-let [regex-result (re-matches regex param-value)]
+      (into {}
+             (mapcat expand-parser-groups group-labels (rest regex-result))))))
+
+;; Functions that use `relative-date-param-values` should feed the `:parser` return
+;; value as the first parameter to both `:range` and `:filter`
+
+;; Note that the order of values in the arrays matter: methods that test parsers against a parameter value
+;; traverse the array from up to bottom
+(def ^:private relative-date-param-values
+  [{:parser  #(= % "today")
+    :range  (fn [_ dt]
+              {:end   dt,
+               :start dt})
+    :filter (fn [_ field] ["=" field ["relative_datetime" "current"]])}
+
+   {:parser  #(= % "yesterday")
+    :range  (fn [_ dt]
+              {:end   (t/minus dt (t/days 1))
+               :start (t/minus dt (t/days 1))})
+    :filter (fn [_ field] ["=" field ["relative_datetime" -1 "day"]])}
+
+   {:parser (regex->parser #"past([0-9]+)(day|week|month|year)s", [:int-value :unit])
+    :range  (fn [{:keys [unit int-value unit-range to-period]} dt]
+              (unit-range (t/minus dt (to-period 1))
+                     (t/minus dt (to-period int-value))))
+    :filter (fn [{:keys [unit int-value]} field]
+              ["TIME_INTERVAL" field (- int-value) unit])}
+
+   {:parser (regex->parser #"next([0-9]+)(day|week|month|year)s" [:int-value :unit])
+    :range  (fn [{:keys [unit int-value unit-range to-period]} dt]
+              (unit-range (t/plus dt (to-period int-value))
+                     (t/plus dt (to-period 1))))
+    :filter (fn [{:keys [unit int-value]} field]
+              ["TIME_INTERVAL" field int-value unit])}
+
+   {:parser (regex->parser #"last(day|week|month|year)" [:unit])
+    :range  (fn [{:keys [unit-range to-period]} dt]
+              (let [last-unit (t/minus dt (to-period 1))]
+                (unit-range last-unit last-unit)))
+    :filter (fn [{:keys [unit]} field]
+              ["TIME_INTERVAL" field "las" unit])}
+
+   {:parser (regex->parser #"this(day|week|month|year)" [:unit])
+    :range  (fn [{:keys [unit-range]} dt]
+              (unit-range dt dt))
+    :filter (fn [{:keys [unit]} field]
+              ["TIME_INTERVAL" field "current" unit])}])
+
+(defn ^:private date->iso8601
+  [date]
```

Comment:
If a function has no comment you should generally put the argslist on the same line as the fn name, e.g.

```clojure
(defn- date->iso8601 [date] 
  ...)
```